crypto/sha/keccak1600.c

```diff
@@ -0,0 +1,281 @@
+                         (uint64_t)inp[4] << 32 | (uint64_t)inp[5] << 40 |
+                         (uint64_t)inp[6] << 48 | (uint64_t)inp[7] << 56;
+            inp += 8;
+        }
+        KeccakF1600(A);
+        len -= r;
+    }
+
+    return len;
+}
+
+/*
+ * SHA3_squeeze is called once at the end to generate |out| hash value
+ * of |len| bytes.
+ */
+void SHA3_squeeze(uint64_t A[5][5], unsigned char *out, size_t len, size_t r)
+{
+    uint64_t *A_flat = (uint64_t *)A;
+    size_t i, rem, w = r / 8;
+
+    while (len >= r) {
+        for (i = 0; i < w; i++) {
+            uint64_t Ai = A_flat[i];
+
+            out[0] = (unsigned char)(Ai);
+            out[1] = (unsigned char)(Ai >> 8);
+            out[2] = (unsigned char)(Ai >> 16);
+            out[3] = (unsigned char)(Ai >> 24);
+            out[4] = (unsigned char)(Ai >> 32);
+            out[5] = (unsigned char)(Ai >> 40);
+            out[6] = (unsigned char)(Ai >> 48);
+            out[7] = (unsigned char)(Ai >> 56);
+            out += 8;
+        }
+        len -= r;
+        if (len)
+            KeccakF1600(A);
+    }
+
+    rem = len % 8;
+    len /= 8;
+
+    for (i = 0; i < len; i++) {
+        uint64_t Ai = A_flat[i];
+
+        out[0] = (unsigned char)(Ai);
+        out[1] = (unsigned char)(Ai >> 8);
+        out[2] = (unsigned char)(Ai >> 16);
+        out[3] = (unsigned char)(Ai >> 24);
+        out[4] = (unsigned char)(Ai >> 32);
+        out[5] = (unsigned char)(Ai >> 40);
+        out[6] = (unsigned char)(Ai >> 48);
+        out[7] = (unsigned char)(Ai >> 56);
+        out += 8;
+    }
+
+    if (rem) {
+        uint64_t Ai = A_flat[i];
+
+        for (i = 0; i < rem; i++) {
+            *out++ = (unsigned char)Ai;
+            Ai >>= 8;
+        }
+    }
+}
+
+#ifdef SELFTEST
+/*
+ * Post-padding one-shot implementations would look as following:
+ *
+ * SHA3_224     SHA3_sponge(inp, len, out, 224/8, (1600-448)/8);
+ * SHA3_256     SHA3_sponge(inp, len, out, 256/8, (1600-512)/8);
+ * SHA3_384     SHA3_sponge(inp, len, out, 384/8, (1600-768)/8);
+ * SHA3_512     SHA3_sponge(inp, len, out, 512/8, (1600-1024)/8);
+ * SHAKE_128    SHA3_sponge(inp, len, out, d, (1600-256)/8);
+ * SHAKE_256    SHA3_sponge(inp, len, out, d, (1600-512)/8);
+ */
+
+void SHA3_sponge(const unsigned char *inp, size_t len,
+                 unsigned char *out, size_t d, size_t r)
+{
+    uint64_t A[5][5];
+
+    memset(A, 0, sizeof(A));
+    SHA3_absorb(A, inp, len, r);
+    SHA3_squeeze(A, out, d, r);
+}
+
+# include <stdio.h>
+
+int main()
+{
+    unsigned char test[168] = { '\xf3', '\x3' };
+    unsigned char out[512];
+    size_t i;
+
+    /*
+     * This is 5-bit SHAKE128 test from http://csrc.nist.gov/groups/ST/toolkit/examples.html#aHashing
+     */
+    test[167] = '\x80';
+    SHA3_sponge(test, sizeof(test), out, sizeof(out), sizeof(test));
+
+    for (i = 0; i < sizeof(out);) {
+        printf("%02X", out[i]);
+        printf(++i % 16 && i != sizeof(out) ? " " : "\n");
+    }
```

Comment:
> how often do you double check evptest.txt.

Trick question, but every chance I get. We are talking about modifications, rather additions, of course. Trick is because it's not always possible to double-check everything, but I'd double-check at least one vector and argue that it "seeds" remaining ones. And here we have one first test :-)